core/modules/user/user.module

```diff
@@ -1044,6 +1060,117 @@ function user_account_form_validate($form, &$form_state) {
+ */
+function user_password_constraints_validate($form, &$form_state) {
+  $strength = 0;
+  if (!array_key_exists('pass', $form_state['values'])) {
+    // If mail verification is required, user registration forms don't have a
+    // password field.
+    return;
+  }
+  $password = trim($form_state['values']['pass']);
+
+  if ($form['#form_id'] == 'user_profile_form' && strlen($password) == 0) {
+    // On profile form and no new password has been set.
+    return;
+  }
+  $config = config('system.core');
+  $messages = array();
+  if ($form['#form_id'] == 'user_pass_reset_form') {
+    $account = $form['#account'];
+    $username = $account->name;
+    $email = $account->mail;
+  }
+  else {
+    $username = $form_state['values']['name'];
+    $email = $form_state['values']['mail'];
+  }
+
+  $strength = _user_estimate_password_entropy($password);
+  if ($strength <= 50) {
+    $messages[] = t('The password is too weak. Please consider making your password more complex.');
+  }
+
+  $min_length_constraint = $config->get('user_password_min_length_constraint');
+  $min_length = $config->get('user_password_min_length');
+  // We need to match the JS string length method which uses UTF-16.
+  if ($min_length_constraint && (mb_strlen($password, 'UTF-16') < $min_length)) {
+    $strength = 5;
+    $messages[] = t('The password is too short. At least @min_length_length characters required.', array('@min_length' => $min_length));
+  }
+  if (backdrop_strtolower($password) == backdrop_strtolower($username)) {
+    $strength = 5;
+    $messages[] = t('The password cannot be the same as the username');
+  }
+  if (backdrop_strtolower($password) == backdrop_strtolower($email)) {
+    $strength = 5;
+    $messages[] = t('The password cannot be the same as the email');
+  }
+
+  if ($strength <= 50) {
+    $form_error = implode('<br>', $messages);
+    form_set_error('pass', $form_error);
+  }
+}
+
+/**
+ * Generates help text for core password constraints on forms.
+ *
+ * @see user_account_form()
+ */
+function _user_password_constraints_help() {
+  $config = config('system.core');
+  $min_length_constraint = $config->get('user_password_min_length_constraint');
+  $min_length = $config->get('user_password_min_length');
+  $help = array();
+  if ($min_length_constraint) {
+    $help[] = t('The password must be at least @min_length characters long.', array('@min_length' => $min_length));
+  }
+  $help[] = t('The password cannot be the same as the username or e-mail address.');
+  $variables = array(
+    'items' => $help,
+    'title' => '',
+    'type' => 'ul',
+    'attributes' => array(),
+  );
+  return theme_item_list($variables);
+}
+
+/**
+ * Estimate password entropy
+ * Adapted from https://github.com/dropbox/zxcvbn.
+ *
+ * @see Backdrop.evaluatePasswordStrength
+ *
+ * @param string $password
+ *
+ * @return int
+ *   An integer representing password strength.
+ */
+function _user_estimate_password_entropy($password) {
```

Comment:
Really nice 👍

Reply:
Thank you. It's functionally the same as the code in user.js and here so we can use it when validating. I'm not sure if this is the best approach - eventually we might want to have the two combined in PHP and have the indicator check the strength via ajax. But for now this is probably fine.